x-pack/plugins/apm/server/lib/helpers/create_es_client/create_apm_event_client/index.ts

```diff
@@ -0,0 +1,128 @@
+  ESSearchResponse,
+} from '../../../../../typings/elasticsearch';
+import {
+  ApmIndicesConfig,
+  ApmIndicesName,
+} from '../../../settings/apm_indices/get_apm_indices';
+import { APMRequestHandlerContext } from '../../../../routes/typings';
+import { addFilterForLegacyData } from './add_filter_for_legacy_data';
+import { callClientWithDebug } from '../call_client_with_debug';
+import { Transaction } from '../../../../../typings/es_schemas/ui/transaction';
+import { Span } from '../../../../../typings/es_schemas/ui/span';
+
+export type APMEventESSearchRequest = Omit<ESSearchRequest, 'index'> & {
+  apm: {
+    events: ProcessorEvent[];
+  };
+};
+
+export const processorEventIndexMap: Record<ProcessorEvent, ApmIndicesName> = {
+  [ProcessorEvent.transaction]: 'apm_oss.transactionIndices',
+  [ProcessorEvent.span]: 'apm_oss.spanIndices',
+  [ProcessorEvent.metric]: 'apm_oss.metricsIndices',
+  [ProcessorEvent.error]: 'apm_oss.errorIndices',
+  [ProcessorEvent.sourcemap]: 'apm_oss.sourcemapIndices',
+  [ProcessorEvent.onboarding]: 'apm_oss.onboardingIndices',
+};
+
+type TypeOfProcessorEvent<T extends ProcessorEvent> = {
+  [ProcessorEvent.error]: APMError;
+  [ProcessorEvent.transaction]: Transaction;
+  [ProcessorEvent.span]: Span;
+  [ProcessorEvent.metric]: APMBaseDoc;
+  [ProcessorEvent.onboarding]: unknown;
+  [ProcessorEvent.sourcemap]: unknown;
+}[T];
+
+type ESSearchRequestOf<TParams extends APMEventESSearchRequest> = Omit<
+  TParams,
+  'apm'
+> & { index: string[] | string };
+
+type TypedSearchResponse<
+  TParams extends APMEventESSearchRequest
+> = ESSearchResponse<
+  TypeOfProcessorEvent<ValuesType<TParams['apm']['events']>>,
+  ESSearchRequestOf<TParams>
+>;
+
+export type APMEventClient = ReturnType<typeof createApmEventClient>;
+
+export function createApmEventClient({
+  context,
+  request,
+  indices,
+  options: { includeFrozen } = { includeFrozen: false },
+}: {
+  context: APMRequestHandlerContext;
+  request: KibanaRequest;
+  indices: ApmIndicesConfig;
+  options: {
+    includeFrozen: boolean;
+  };
+}) {
+  const client = context.core.elasticsearch.legacy.client;
+
+  return {
+    search<TParams extends APMEventESSearchRequest>(
+      params: TParams,
+      { includeLegacyData } = { includeLegacyData: false }
+    ): Promise<TypedSearchResponse<TParams>> {
+      const { apm, ...esParams } = params;
+
+      const index = uniq(
+        apm.events.map((event) => indices[processorEventIndexMap[event]])
+      );
+
+      const nextParams: ESSearchRequest & {
+        body: { query: { bool: { filter: ESFilter[] } } };
+      } = defaultsDeep(cloneDeep(esParams), {
+        body: {
+          query: {
+            bool: {
+              filter: [],
+            },
+          },
+        },
+      });
+
+      nextParams.body.query.bool.filter.push({
+        terms: {
+          [PROCESSOR_EVENT]: apm.events,
+        },
+      });
+
+      const withLegacyData = addFilterForLegacyData(nextParams, {
```

Comment:
nit: "withLegacyData" is a little misleading since it's a filter to exclude legacy data (but it can be disabled)

```suggestion
      const paramsWithLegacyFilter = addFilterForLegacyData(nextParams, {
```